packages/core/src/common/message-rpc/rpc-protocol.ts

```diff
@@ -45,6 +45,31 @@ export interface RpcProtocolOptions {
     mode?: 'default' | 'clientOnly' | 'serverOnly'
 }
 
+/**
+ * Wrapper for a {@link Disposable} that is not available immediately.
+ */
+export class MaybeDisposable {
```

Comment:
I really like the idea of this class and I believe we can even add it to the common `disposable.ts`. From my understanding it is a wrapper that you mark as disposed and either the already given disposable or any later set disposable is then also disposed, making sure that whatever is given to the wrapper has the correct disposed state.

So I think that this class should actually also implement `Disposable` so it actually acts more like a wrapper fulfilling the same interface. Also, from the name alone it is not very clear what this class does, so maybe something like `DelayedDisposable` or actually `WrappedDisposable` or `DisposableWrapper` might be better. I believe we could even see this as a special case of the `DisposableCollection` with an adapted `push` that does not add things to the collection once dispose was called but instead simply disposes anything that is added to it, so maybe we could also adapt it to that. That would also solve the problem of people calling `setDisposable` multiple times without knowing whether we already wrapped a disposable or not.

However, I'll leave it up to you, I just think we should share it in a more common API and fix the naming a bit.

Reply:
I'll move it to `disposable.ts`. I am not sure about `DisposableCollection` though. If something like this is needed with a collection I would rather expect people placing the wrapper inside a disposable collection, which is possible now with implementing `Disposable`.

Reply:
Sure, I'm fine with that.